### ISSUE TITLE
Interpolations in CJK and Cyrillic languages broken in IE11

### ISSUE THREAD

**Issue by jshoudy11** [angular/angular.js#14924]
**Bug**
Certain interpolations are being broken into multiple spans. For example in a template we have
\<span>{{ctrl.variableName}}\</span>.

In Chrome, Firefox, Safari, Edge this always becomes \<span>Text associated with variable\</span>

but in IE11 in CJK languages this is becoming \<span>{{\</span>\<span>ctrl.variableName}}\</span> which displays on our page as the variable.

There are several places this bug is displayed in production for our site but I will include steps to reproduce for just one of them.

**To reproduce:**

1) Go to https://console.cloud.google.com in IE11
2) Click the hamburger menu in the top left -> click API Manager
3) Click Credentials in the left side nav (if you're prompted to create a project click "Create")
4) Click "Create credentials" -> "API Key" -> "Server Key" -> "Create"
5) Go back to the Credentials Page and click on the name of your new API Key
6) Click the vertical "..." in the top right -> Preferences -> Language and regional formats
7) Change the language to a CJK language (Japanese for example) and click Save
8) The page will reload. Observe the {{ctrl.key.currentKey}} and other variables displayed in the page. 

Here is a screenshot of the issue: 
![angularbug](https://cloud.githubusercontent.com/assets/17074023/16890979/c2b38718-4aa8-11e6-8ed5-18511dc328bd.PNG)

** Details **
This is a bug specific to IE11 (and earlier IE). I'm using IE version 11.0.10240.16603 on Windows 10 Enterprise. We are running angular 1.5 but this also occurred in 1.4

**Comment by gkalpak** [comment]
I can reproduce this. A very similar quirk was fixed with #11796. Ooc, are you using MutationObservers?

**Comment by gkalpak** [comment]
You don't happen to have a more minimal reproduction (or at least one with unminified sources), do you? :stuck_out_tongue:

**Comment by gkalpak** [comment]
I tried my best to reproduce this, but failed miserably. And I have been equally unsuccessful at debugging this directly on https://console.cloud.google.com in IE11. :worried: 

If someone is able to provide a more "debuggable" reproduction or has access to investigate on the dev environment of https://console.cloud.google.com, that would help a lot :smile:

**Comment by jshoudy11** [comment]
Hey Georgios, thanks for looking into this. I have also not had much luck with getting a minimal reproduction but I do have access to the dev environment. Do you have any ideas for places where I can start to debug?

**Comment by gkalpak** [comment]
@jshoudy11 , I was finally able to reproduce it :tada: It seems to be a combination of `ngTransclude` with the known IE11 bug wrt to MutationObservers and `{{` characters following "special characters", such as `&mdash;` (or CJK characters).

It is interesting that there is different behavior between 1.5.7 and the current master (probably as a result of our stopping to wrap transcluded text-content in spans (as of 7617c08). Just to be clear, the behavior is problematic on both versions, but different: [1.5.7 demo][1], [snapshot demo][2]

(Now, all we need to find out is whether/how we can work around this :smile:)

[1.5.7]: http://codepen.io/ExpertSystem/pen/zBRjmj?editors=1001
[snapshot]: http://codepen.io/ExpertSystem/pen/JKpkPx?editors=1001

**Comment by gkalpak** [comment]
Btw, the simplest work-around for application code, seems to be wrapping your transcluded text nodes in spans. E.g.:

```html
<!-- BEFORE: -->
<directive-with-transclusion>
  My text content
</directive-with-transclusion>

<!-- AFTER: -->
<directive-with-transclusion>
  <span>My text content</span>
</directive-with-transclusion>
```

**Comment by lgalfaso** [comment]
This is a wild guess on what the issue is (as I do not have access to IE). First IE by using MutationObservers splits the text into two nodes. This is common for 1.5 and 1.6

In 1.5 the issue manifest like this, here
https://github.com/angular/angular.js/blob/v1.5.x/src/ng/compile.js#L1758
we do not look for two consecutive text nodes, then these are wrapped in span element, so the interpolation fails.

In 1.6 here
https://github.com/angular/angular.js/blob/master/src/ng/compile.js#L2047
given that the content is transcluded, there is no parent and the nodes are not merged.

**Comment by jshoudy11** [comment]
Thank you for the help @gkalpak and @lgalfaso. I actually do not believe we are using MutationObservers in the areas where this error is occurring. But I will try to wrap the interpolations in question in span tags and see if it works. I don't have access to IE today but I will tomorrow. I'll let you know if it works.

**Comment by Narretz** [comment]
@gkalpak your snapshot demo is broken in IE11 for me even without mutation observers.

**Comment by gkalpak** [comment]
@Narretz , IE apparently has this "feature" that if you run it once with the MutationObserver, removing it and clicking the "Run" button doesn't fix the problem (it is as if it is running with the MutationObserver).

You have to save it (e.g. fork the pen) without the MutationObserver. Only when reloading the whole page do changes in MutationObserver take effect (or something along these lines).

**Comment by Narretz** [comment]
Ah, okay, thanks for the explanation. IE strikes again.

**Comment by jshoudy11** [comment]
@gkalpak Confirmed that wrapping in span tags is a functioning workaround for us. Thank you for the help. I'm sure you guys will keep working on a more permanent workaround within angular and I'll stay tuned.

**Comment by jshoudy11** [comment]
@gkalpakis is there an ETA on a solution to this in the angular code?

**Comment by gkalpak** [comment]
@jshoudy11, I am planning to look into this today. So, (if all goes well) the fix will be in 1.5.9.

**Comment by jshoudy11** [comment]
@gkalpak what is the status of this?

**Comment by gkalpak** [comment]
@jshoudy11, I have a fix for this on my fork (for 1.6.x), but I am not very happy with it. And it's kind of a breaking change, so I guess it won't be backported to 1.5.x. Here is the fix: https://github.com/gkalpak/angular.js/commit/3ad0acddd0f3553c5698105ecab63469dd3dc1f7

Your safest bet is still to always wrap transcuded (top-level) text nodes. I'll open a PR soon.

**Comment by gkalpak** [comment]
@jshoudy11, #15025 has been merged. Feel free to try it out and report back :smile:

**Comment by jshoudy11** [comment]
@gkalpak is it only in 1.6 (as you originally suspected) or is it backported to 1.5.x as well?

**Comment by gkalpak** [comment]
1.5.x only.

**Comment by jshoudy11** [comment]
@gkalpak will this ever be backported to 1.5.x? Our project is currently on 1.5 with no immediate plans to upgrade to 1.6. We have to continue to use the span work around unless this is backported

**Comment by jshoudy11** [comment]
Other question: when will 1.6 be available in google3? Currently 1.5 is the latest version

**Comment by gkalpak** [comment]
> @gkalpak will this ever be backported to 1.5.x?

Nope, because it is a breaking change.

> Our project is currently on 1.5 with no immediate plans to upgrade to 1.6. We have to continue to use the span work around unless this is backported

Iirc you were using custom snapshot build, no? If that is true, you are already using 1.6 pre-alpha :smiley:

> Other question: when will 1.6 be available in google3? Currently 1.5 is the latest version

It has to be released first, I guess (which will hopefully happen soon(-ish)).
